Sources/SKCore/BuildSystemManager.swift

```diff
@@ -181,35 +183,31 @@ extension BuildSystemManager {
     }
   }
 
-  public func registerForChangeNotifications(for uri: DocumentURI, language: Language) {
-    return queue.async {
-      log("registerForChangeNotifications(\(uri.pseudoPath))")
-      let mainFile: DocumentURI
+  public func registerForChangeNotifications(for uri: DocumentURI, language: Language) async {
+    log("registerForChangeNotifications(\(uri.pseudoPath))")
+    let mainFile: DocumentURI
 
-      if let watchedFile = self.watchedFiles[uri] {
-        mainFile = watchedFile.mainFile
-      } else {
-        let mainFiles = self._mainFilesProvider?.mainFilesContainingFile(uri)
-        mainFile = chooseMainFile(for: uri, from: mainFiles ?? [])
-        self.watchedFiles[uri] = (mainFile, language)
-      }
+    if let watchedFile = self.watchedFiles[uri] {
+      mainFile = watchedFile.mainFile
+    } else {
+      let mainFiles = self._mainFilesProvider?.mainFilesContainingFile(uri)
+      mainFile = chooseMainFile(for: uri, from: mainFiles ?? [])
+      self.watchedFiles[uri] = (mainFile, language)
+    }
 
-      let newStatus = self.cachedStatusOrRegisterForSettings(for: mainFile, language: language)
+    let newStatus = await self.cachedStatusOrRegisterForSettings(for: mainFile, language: language)
 
-      if let mainChange = newStatus.buildSettingsChange,
-         let delegate = self._delegate {
-        let change = self.convert(change: mainChange, ofMainFile: mainFile, to: uri)
-        self.notifyQueue.async {
-          delegate.fileBuildSettingsChanged([uri: change])
-        }
-      }
+    if let mainChange = newStatus.buildSettingsChange,
+       let delegate = self._delegate {
+      let change = self.convert(change: mainChange, ofMainFile: mainFile, to: uri)
+      await delegate.fileBuildSettingsChanged([uri: change])
```

Comment:
Yes, that’s intentional. I checked that none of these methods are doing a huge amount of work. I decided to await their results here because otherwise we cannot guarantee that the callbacks will be called in order (if tasks aren’t executed in the order they were created).

Reply:
Makes sense, thanks